CounterpointGenerator/CounterpointGenerator/Rules/BattutaRule.cs

```diff
@@ -13,27 +13,37 @@ class BattutaRule : IRules
         /**
          * INPUTS: Possibilities, CurrentNote, PreviousCantus, PreviousCounterpoint
          */
-        public List<int> Apply(RuleInput ruleInput)
+        public List<Note> Apply(RuleInput ruleInput)
         {
-            List<int> output = new List<int>(ruleInput.Possibilities);
+            // Do nothing if no previous notes!
+            if (ruleInput.Position == 0)
+            {
+                return ruleInput.Possibilities;
+            }
+
+            List<Note> output = new List<Note>(ruleInput.Possibilities);
             // Check if previous cantus stepped up
-            if (ruleInput.PreviousCantus < ruleInput.CurrentNote)
+            if (ruleInput.PreviousCantus.Pitch < ruleInput.CurrentNote.Pitch)
             {
                 // Check if previous counterpoint was the higher voice
-                if(ruleInput.PreviousCounterpoint > ruleInput.PreviousCantus)
+                if(ruleInput.PreviousCounterpoint.Pitch > ruleInput.PreviousCantus.Pitch)
                 {
                     // Remove upper octave as a possibility
-                    output.Remove(ruleInput.CurrentNote + 13);
+                    Note removeNote = new Note(ruleInput.CurrentNote.Pitch + 13, ruleInput.CurrentNote.Length);
```

Comment:
It looks like you meant to remove an octave of possibilities, not just the note an octave above this note. Perhaps you want output.RemoveAll(n => n.Pitch > ruleInput.CurrentNote.Pitch);

Reply:
No, I meant to specifically remove only the octave here. If the higher voice steps down and the lower voice steps up they cannot be an octave apart. You can see the original version before the change had only to remove the int value equal to an octave, the first thing I typed had Remove(new Note(/*octave*/)) which I later realized wouldn't work because even though the values are the same the object is different